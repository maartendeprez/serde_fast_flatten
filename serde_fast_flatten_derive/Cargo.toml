[package]
name = "serde_fast_flatten_derive"
description = "A faster `flatten` implementation for serde"
authors = ["Maarten Deprez <deprez.maarten@gmail.com>"]
keywords = ["serde", "serialization", "flatten"]
categories = ["encoding"]

edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
version.workspace = true
# rust-version.workspace = true

[lib]
proc-macro = true

[features]
default = []

[dependencies]
proc-macro2 = "1.0.93"
quote = "1.0.38"
serde_derive_internals = "0.29.1"
syn = "2.0.98"
# prettyplease = "0.2.29"

[dev-dependencies]
bincode = "1.3.3"
bitcode = { version = "0.6.4", features = ["serde"] }
insta = "1.42.1"
serde = "1.0.217"
serde_cbor = "0.11.2"
serde_fast_flatten = { version = "0.1.0", path = "../serde_fast_flatten", features = [
    "derive",
] }
serde_json = "1.0.138"
serde_yaml = "0.9.34"
