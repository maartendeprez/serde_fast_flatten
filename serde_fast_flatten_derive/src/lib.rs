extern crate proc_macro;

use proc_macro::TokenStream;
use syn::{parse_macro_input, DeriveInput};

mod deserialize_fields;
mod serialize_fields;

#[proc_macro_derive(SerializeFields, attributes(serde))]
pub fn derive_serialize_fields(item: TokenStream) -> TokenStream {
    let input: DeriveInput = parse_macro_input!(item);
    let tokens = serialize_fields::serialize_fields(input);
    // eprintln!(
    //     "{}",
    //     prettyplease::unparse(&syn::parse_file(&tokens.to_string()).unwrap())
    // );
    TokenStream::from(tokens)
}

#[proc_macro_derive(DeserializeFields, attributes(serde))]
pub fn derive_deserialize_fields(item: TokenStream) -> TokenStream {
    let input: DeriveInput = parse_macro_input!(item);
    let tokens = deserialize_fields::deserialize_fields(input);
    // eprintln!(
    //     "{}",
    //     prettyplease::unparse(&syn::parse_file(&tokens.to_string()).unwrap())
    // );
    TokenStream::from(tokens)
}
