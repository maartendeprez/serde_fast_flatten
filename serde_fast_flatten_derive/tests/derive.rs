use insta::{assert_binary_snapshot, assert_snapshot};
use serde_fast_flatten_derive::{DeserializeFields, SerializeFields};

#[derive(SerializeFields, DeserializeFields, PartialEq, Eq, Debug)]
struct Address {
    #[serde(flatten)]
    person: Person,
    #[serde(flatten)]
    street_address: StreetAddress,
}

#[derive(SerializeFields, DeserializeFields, PartialEq, Eq, Debug)]
struct Person {
    first_name: String,
    last_name: String,
}

#[derive(SerializeFields, DeserializeFields, PartialEq, Eq, Debug)]
struct StreetAddress {
    street: String,
    house_number: u64,
    #[serde(flatten)]
    city_address: CityAddress,
}

#[derive(SerializeFields, DeserializeFields, PartialEq, Eq, Debug)]
struct CityAddress {
    city: String,
    postal_code: String,
    state: Option<String>,
    country: String,
}

fn test_value() -> Address {
    Address {
        person: Person {
            first_name: String::from("John"),
            last_name: String::from("Lennon"),
        },
        street_address: StreetAddress {
            street: String::from("Long and Winding Road"),
            house_number: 42,
            city_address: CityAddress {
                city: String::from("Strawberry Fields"),
                postal_code: String::from("1234"),
                state: None,
                country: String::from("Nowhere Land"),
            },
        },
    }
}

#[test]
fn roundtrip_json() {
    let x = test_value();
    let s = serde_json::to_string(&x).unwrap();
    let y = serde_json::from_str(&s).unwrap();
    assert_eq!(x, y);
    assert_snapshot!(s);
}

#[test]
fn roundtrip_yaml() {
    let x = test_value();
    let s = serde_yaml::to_string(&x).unwrap();
    let y = serde_yaml::from_str(&s).unwrap();
    assert_eq!(x, y);
    assert_snapshot!(s);
}

#[test]
fn roundtrip_cbor() {
    let x = test_value();
    let s = serde_cbor::to_vec(&x).unwrap();
    let y = serde_cbor::from_slice(&s).unwrap();
    assert_eq!(x, y);
    assert_binary_snapshot!(".cbor", s);
}

#[test]
fn roundtrip_cbor_packed() {
    let x = test_value();
    let s = serde_cbor::ser::to_vec_packed(&x).unwrap();
    let y = serde_cbor::from_slice(&s).unwrap();
    assert_eq!(x, y);
    assert_binary_snapshot!(".cbor", s);
}

#[test]
fn roundtrip_bincode() {
    let x = test_value();
    let s = bincode::serialize(&x).unwrap();
    let y = bincode::deserialize(&s).unwrap();
    assert_eq!(x, y);
    assert_binary_snapshot!(".bincode", s);
}

#[test]
fn roundtrip_bincode_compact() {
    use bincode::Options;
    let opts = bincode::DefaultOptions::new();
    let x = test_value();
    let s = opts.serialize(&x).unwrap();
    let y = opts.deserialize(&s).unwrap();
    assert_eq!(x, y);
    assert_binary_snapshot!(".bincode", s);
}

#[test]
fn roundtrip_bitcode() {
    let x = test_value();
    let s = bitcode::serialize(&x).unwrap();
    let y = bitcode::deserialize(&s).unwrap();
    assert_eq!(x, y);
    assert_binary_snapshot!(".bitcode", s);
}
