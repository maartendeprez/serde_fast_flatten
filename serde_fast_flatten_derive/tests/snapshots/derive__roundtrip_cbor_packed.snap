---
source: serde_fast_flatten_derive/tests/derive.rs
expression: s
extension: cbor
snapshot_kind: binary
---
